File: src/storage/sqlite.zig
const std = @import("std");
const abi = @import("sqlite_abi.zig");
const model = @import("../model/allocation.zig");

pub const Repository = struct {
    db: *abi.sqlite3,

    pub fn open(_: std.mem.Allocator, path: []const u8) !Repository {
        var db_opt: ?*abi.sqlite3 = null;
        const zpath = try std.heap.c_allocator.dupeZ(u8, path);
        defer std.heap.c_allocator.free(zpath);
        try abi.check(abi.sqlite3_open_v2(zpath.ptr, &db_opt, abi.SQLITE_OPEN_READWRITE | abi.SQLITE_OPEN_CREATE | abi.SQLITE_OPEN_FULLMUTEX, null), null);
        const db = db_opt orelse return error.SqliteFailure;
        errdefer _ = abi.sqlite3_close_v2(db);
        try abi.check(abi.sqlite3_busy_timeout(db, 5000), db);
        try exec(db, "PRAGMA journal_mode=WAL;");
        try exec(db, "CREATE TABLE IF NOT EXISTS allocations (id TEXT PRIMARY KEY, protocol TEXT NOT NULL, port INTEGER NOT NULL, target_port INTEGER NOT NULL, host TEXT, created_at_ms INTEGER NOT NULL, updated_at_ms INTEGER NOT NULL, UNIQUE(protocol, port));");
        try exec(db, "CREATE TABLE IF NOT EXISTS bindings (allocation_id TEXT PRIMARY KEY, target_port INTEGER NOT NULL, host TEXT, created_at_ms INTEGER NOT NULL, updated_at_ms INTEGER NOT NULL);");
        try exec(
            db,
            "INSERT INTO bindings(allocation_id, target_port, host, created_at_ms, updated_at_ms) " ++
                "SELECT id, target_port, host, created_at_ms, updated_at_ms FROM allocations " ++
                "WHERE target_port > 0 AND NOT EXISTS (SELECT 1 FROM bindings WHERE bindings.allocation_id = allocations.id);",
        );
        return .{ .db = db };
    }

    pub fn close(self: *Repository) void {
        _ = abi.sqlite3_close_v2(self.db);
    }

    pub fn selfCheck(self: *Repository) !void {
        var stmt = try prepare(self.db, "SELECT sqlite_version();");
        defer stmt.deinit();
        try abi.check(abi.sqlite3_step(stmt.stmt), self.db);
        _ = abi.sqlite3_column_text(stmt.stmt, 0) orelse return error.SqliteFailure;
    }

    pub fn begin(self: *Repository) !void {
        try exec(self.db, "BEGIN IMMEDIATE TRANSACTION;");
    }

    pub fn commit(self: *Repository) !void {
        try exec(self.db, "COMMIT;");
    }

    pub fn rollback(self: *Repository) void {
        exec(self.db, "ROLLBACK;") catch {};
    }

    pub fn insertAllocation(self: *Repository, allocation: model.Allocation) !void {
        var stmt = try prepare(self.db, "INSERT INTO allocations(id, protocol, port, target_port, host, created_at_ms, updated_at_ms) VALUES(?, ?, ?, ?, ?, ?, ?);");
        defer stmt.deinit();
        try stmt.bindText(1, allocation.id);
        try stmt.bindText(2, allocation.protocol.asString());
        try stmt.bindInt(3, allocation.port);
        try stmt.bindInt(4, allocation.target_port orelse 0);
        try stmt.bindOptionalText(5, allocation.host);
        try stmt.bindInt64(6, allocation.created_at_ms);
        try stmt.bindInt64(7, allocation.updated_at_ms);
        try stmt.done(self.db);
    }

    pub fn putBinding(self: *Repository, binding: model.Binding) !void {
        var stmt = try prepare(
            self.db,
            "INSERT INTO bindings(allocation_id, target_port, host, created_at_ms, updated_at_ms) VALUES(?, ?, ?, ?, ?) " ++
                "ON CONFLICT(allocation_id) DO UPDATE SET target_port = excluded.target_port, host = excluded.host, updated_at_ms = excluded.updated_at_ms;",
        );
        defer stmt.deinit();
        try stmt.bindText(1, binding.allocation_id);
        try stmt.bindInt(2, binding.target_port);
        try stmt.bindOptionalText(3, binding.host);
        try stmt.bindInt64(4, binding.created_at_ms);
        try stmt.bindInt64(5, binding.updated_at_ms);
        try stmt.done(self.db);

        try self.updateLegacyBindingColumns(binding.allocation_id, binding.target_port, binding.host, binding.updated_at_ms);
    }

    pub fn deleteBinding(self: *Repository, allocation_id: []const u8, updated_at_ms: i64) !bool {
        var stmt = try prepare(self.db, "DELETE FROM bindings WHERE allocation_id = ?;");
        defer stmt.deinit();
        try stmt.bindText(1, allocation_id);
        try stmt.done(self.db);
        const changed = abi.sqlite3_changes(self.db) > 0;
        if (changed) {
            try self.clearLegacyBindingColumns(allocation_id, updated_at_ms);
        }
        return changed;
    }

    pub fn deleteAllocation(self: *Repository, id: []const u8) !bool {
        var delete_binding = try prepare(self.db, "DELETE FROM bindings WHERE allocation_id = ?;");
        defer delete_binding.deinit();
        try delete_binding.bindText(1, id);
        try delete_binding.done(self.db);

        var stmt = try prepare(self.db, "DELETE FROM allocations WHERE id = ?;");
        defer stmt.deinit();
        try stmt.bindText(1, id);
        try stmt.done(self.db);
        return abi.sqlite3_changes(self.db) > 0;
    }

    pub fn getBinding(self: *Repository, allocator: std.mem.Allocator, allocation_id: []const u8) !?model.Binding {
        var stmt = try prepare(self.db, "SELECT allocation_id, target_port, host, created_at_ms, updated_at_ms FROM bindings WHERE allocation_id = ?;");
        defer stmt.deinit();
        try stmt.bindText(1, allocation_id);
        const rc = abi.sqlite3_step(stmt.stmt);
        if (rc == abi.SQLITE_DONE) return null;
        try abi.check(rc, self.db);
        return try rowToBinding(allocator, stmt.stmt);
    }

    pub fn getAllocation(self: *Repository, allocator: std.mem.Allocator, id: []const u8) !?model.Allocation {
        var stmt = try prepare(
            self.db,
            "SELECT a.id, a.protocol, a.port, COALESCE(b.target_port, NULLIF(a.target_port, 0)), COALESCE(b.host, a.host), a.created_at_ms, a.updated_at_ms " ++
                "FROM allocations a LEFT JOIN bindings b ON b.allocation_id = a.id WHERE a.id = ?;",
        );
        defer stmt.deinit();
        try stmt.bindText(1, id);
        const rc = abi.sqlite3_step(stmt.stmt);
        if (rc == abi.SQLITE_DONE) return null;
        try abi.check(rc, self.db);
        return try rowToAllocation(allocator, stmt.stmt);
    }

    pub fn listAllocations(self: *Repository, allocator: std.mem.Allocator) !std.ArrayList(model.Allocation) {
        var rows = std.ArrayList(model.Allocation).empty;
        errdefer {
            for (rows.items) |*item| item.deinit(allocator);
            rows.deinit(allocator);
        }
        var stmt = try prepare(
            self.db,
            "SELECT a.id, a.protocol, a.port, COALESCE(b.target_port, NULLIF(a.target_port, 0)), COALESCE(b.host, a.host), a.created_at_ms, a.updated_at_ms " ++
                "FROM allocations a LEFT JOIN bindings b ON b.allocation_id = a.id ORDER BY a.protocol, a.port;",
        );
        defer stmt.deinit();
        while (true) {
            const rc = abi.sqlite3_step(stmt.stmt);
            if (rc == abi.SQLITE_DONE) break;
            try abi.check(rc, self.db);
            try rows.append(allocator, try rowToAllocation(allocator, stmt.stmt));
        }
        return rows;
    }

    fn updateLegacyBindingColumns(self: *Repository, allocation_id: []const u8, target_port: u16, host: ?[]const u8, updated_at_ms: i64) !void {
        var stmt = try prepare(self.db, "UPDATE allocations SET target_port = ?, host = ?, updated_at_ms = ? WHERE id = ?;");
        defer stmt.deinit();
        try stmt.bindInt(1, target_port);
        try stmt.bindOptionalText(2, host);
        try stmt.bindInt64(3, updated_at_ms);
        try stmt.bindText(4, allocation_id);
        try stmt.done(self.db);
    }

    fn clearLegacyBindingColumns(self: *Repository, allocation_id: []const u8, updated_at_ms: i64) !void {
        var stmt = try prepare(self.db, "UPDATE allocations SET target_port = 0, host = NULL, updated_at_ms = ? WHERE id = ?;");
        defer stmt.deinit();
        try stmt.bindInt64(1, updated_at_ms);
        try stmt.bindText(2, allocation_id);
        try stmt.done(self.db);
    }
};

const Statement = struct {
    stmt: *abi.sqlite3_stmt,

    fn deinit(self: *Statement) void {
        _ = abi.sqlite3_finalize(self.stmt);
    }

    fn bindText(self: *Statement, idx: c_int, text: []const u8) !void {
        try abi.check(abi.sqlite3_bind_text(self.stmt, idx, text.ptr, @intCast(text.len), abi.SQLITE_TRANSIENT), null);
    }

    fn bindOptionalText(self: *Statement, idx: c_int, text: ?[]const u8) !void {
        if (text) |v| try self.bindText(idx, v) else try self.bindNull(idx);
    }

    fn bindInt(self: *Statement, idx: c_int, value: anytype) !void {
        try abi.check(abi.sqlite3_bind_int(self.stmt, idx, @intCast(value)), null);
    }

    fn bindInt64(self: *Statement, idx: c_int, value: i64) !void {
        try abi.check(abi.sqlite3_bind_int64(self.stmt, idx, value), null);
    }

    fn bindNull(self: *Statement, idx: c_int) !void {
        try abi.check(abi.sqlite3_bind_null(self.stmt, idx), null);
    }

    fn done(self: *Statement, db: *abi.sqlite3) !void {
        try abi.check(abi.sqlite3_step(self.stmt), db);
    }
};

fn exec(db: *abi.sqlite3, sql: []const u8) !void {
    const zsql = try std.heap.c_allocator.dupeZ(u8, sql);
    defer std.heap.c_allocator.free(zsql);
    try abi.check(abi.sqlite3_exec(db, zsql.ptr, null, null, null), db);
}

fn prepare(db: *abi.sqlite3, sql: []const u8) !Statement {
    const zsql = try std.heap.c_allocator.dupeZ(u8, sql);
    defer std.heap.c_allocator.free(zsql);
    var stmt_opt: ?*abi.sqlite3_stmt = null;
    try abi.check(abi.sqlite3_prepare_v2(db, zsql.ptr, -1, &stmt_opt, null), db);
    return .{ .stmt = stmt_opt orelse return error.SqliteFailure };
}

fn rowToAllocation(allocator: std.mem.Allocator, stmt: *abi.sqlite3_stmt) !model.Allocation {
    const id = try allocator.dupe(u8, std.mem.span(abi.sqlite3_column_text(stmt, 0).?));
    errdefer allocator.free(id);
    const host_ptr = abi.sqlite3_column_text(stmt, 4);
    return .{
        .id = id,
        .protocol = model.Protocol.fromString(std.mem.span(abi.sqlite3_column_text(stmt, 1).?)) orelse return error.InvalidData,
        .port = @intCast(abi.sqlite3_column_int(stmt, 2)),
        .target_port = if (abi.sqlite3_column_type(stmt, 3) == abi.SQLITE_NULL) null else blk: {
            const value = abi.sqlite3_column_int(stmt, 3);
            if (value <= 0) break :blk null;
            break :blk @as(u16, @intCast(value));
        },
        .host = if (host_ptr) |ptr| try allocator.dupe(u8, std.mem.span(ptr)) else null,
        .created_at_ms = abi.sqlite3_column_int64(stmt, 5),
        .updated_at_ms = abi.sqlite3_column_int64(stmt, 6),
    };
}

fn rowToBinding(allocator: std.mem.Allocator, stmt: *abi.sqlite3_stmt) !model.Binding {
    const allocation_id = try allocator.dupe(u8, std.mem.span(abi.sqlite3_column_text(stmt, 0).?));
    errdefer allocator.free(allocation_id);
    const host_ptr = abi.sqlite3_column_text(stmt, 2);
    return .{
        .allocation_id = allocation_id,
        .target_port = @intCast(abi.sqlite3_column_int(stmt, 1)),
        .host = if (host_ptr) |ptr| try allocator.dupe(u8, std.mem.span(ptr)) else null,
        .created_at_ms = abi.sqlite3_column_int64(stmt, 3),
        .updated_at_ms = abi.sqlite3_column_int64(stmt, 4),
    };
}

File: src/storage/sqlite.rs
use crate::model::{Allocation, Binding, Protocol};
use sea_orm::{
    ActiveModelTrait, ActiveValue::Set, ConnectionTrait, DatabaseConnection, DbBackend, DbErr,
    EntityTrait, IntoActiveModel, QueryOrder, Schema, Statement,
};
use sqlx::SqlitePool;
use sqlx::sqlite::{SqliteConnectOptions, SqliteJournalMode, SqlitePoolOptions};
use std::collections::HashMap;
use std::num::TryFromIntError;
use std::str::FromStr;
use std::time::Duration;

mod allocation_entity {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
    #[sea_orm(table_name = "allocations")]
    pub struct Model {
        #[sea_orm(primary_key, auto_increment = false)]
        pub id: String,
        pub protocol: String,
        pub port: i32,
        pub target_port: i32,
        pub host: Option<String>,
        pub created_at_ms: i64,
        pub updated_at_ms: i64,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

mod binding_entity {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
    #[sea_orm(table_name = "bindings")]
    pub struct Model {
        #[sea_orm(primary_key, auto_increment = false)]
        pub allocation_id: String,
        pub target_port: i32,
        pub host: Option<String>,
        pub created_at_ms: i64,
        pub updated_at_ms: i64,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

pub struct Repository {
    pool: SqlitePool,
    db: DatabaseConnection,
}

#[derive(Debug, thiserror::Error)]
pub enum RepositoryError {
    #[error(transparent)]
    Db(#[from] DbErr),
    #[error(transparent)]
    Sqlx(#[from] sqlx::Error),
    #[error(transparent)]
    Model(#[from] crate::model::ModelError),
    #[error("invalid {field} value in database: {value}")]
    InvalidPort { field: &'static str, value: i32 },
    #[error(transparent)]
    PortConversion(#[from] TryFromIntError),
}

pub type Result<T> = std::result::Result<T, RepositoryError>;

impl Repository {
    pub async fn open(path: impl AsRef<std::path::Path>) -> Result<Self> {
        let options = SqliteConnectOptions::new()
            .filename(path)
            .create_if_missing(true)
            .journal_mode(SqliteJournalMode::Wal)
            .busy_timeout(Duration::from_millis(5000));
        let pool = SqlitePoolOptions::new()
            .max_connections(1)
            .connect_with(options)
            .await?;
        let db = sea_orm::SqlxSqliteConnector::from_sqlx_sqlite_pool(pool.clone());
        let repo = Self { pool, db };
        repo.setup_schema().await?;
        repo.migrate_legacy_bindings().await?;
        Ok(repo)
    }

    pub async fn self_check(&self) -> Result<()> {
        let version: String = sqlx::query_scalar("SELECT sqlite_version();")
            .fetch_one(&self.pool)
            .await?;
        if version.is_empty() {
            return Err(sqlx::Error::RowNotFound.into());
        }
        Ok(())
    }

    pub async fn insert_allocation(&self, allocation: &Allocation) -> Result<()> {
        allocation_entity::ActiveModel {
            id: Set(allocation.id.clone()),
            protocol: Set(allocation.protocol.as_str().to_owned()),
            port: Set(i32::from(allocation.port)),
            target_port: Set(allocation.target_port.map(i32::from).unwrap_or(0)),
            host: Set(allocation.host.clone()),
            created_at_ms: Set(allocation.created_at_ms),
            updated_at_ms: Set(allocation.updated_at_ms),
        }
        .insert(&self.db)
        .await?;
        Ok(())
    }

    pub async fn put_binding(&self, binding: &Binding) -> Result<()> {
        match binding_entity::Entity::find_by_id(binding.allocation_id.clone())
            .one(&self.db)
            .await?
        {
            Some(model) => {
                let mut active = model.into_active_model();
                active.target_port = Set(i32::from(binding.target_port));
                active.host = Set(binding.host.clone());
                active.updated_at_ms = Set(binding.updated_at_ms);
                active.update(&self.db).await?;
            }
            None => {
                binding_entity::ActiveModel {
                    allocation_id: Set(binding.allocation_id.clone()),
                    target_port: Set(i32::from(binding.target_port)),
                    host: Set(binding.host.clone()),
                    created_at_ms: Set(binding.created_at_ms),
                    updated_at_ms: Set(binding.updated_at_ms),
                }
                .insert(&self.db)
                .await?;
            }
        }
        self.update_legacy_binding_columns(
            &binding.allocation_id,
            binding.target_port,
            binding.host.clone(),
            binding.updated_at_ms,
        )
        .await?;
        Ok(())
    }

    pub async fn delete_binding(&self, allocation_id: &str, updated_at_ms: i64) -> Result<bool> {
        let result = binding_entity::Entity::delete_by_id(allocation_id.to_owned())
            .exec(&self.db)
            .await?;
        let changed = result.rows_affected > 0;
        if changed {
            self.clear_legacy_binding_columns(allocation_id, updated_at_ms)
                .await?;
        }
        Ok(changed)
    }

    pub async fn delete_allocation(&self, id: &str) -> Result<bool> {
        binding_entity::Entity::delete_by_id(id.to_owned())
            .exec(&self.db)
            .await?;
        let result = allocation_entity::Entity::delete_by_id(id.to_owned())
            .exec(&self.db)
            .await?;
        Ok(result.rows_affected > 0)
    }

    pub async fn get_binding(&self, allocation_id: &str) -> Result<Option<Binding>> {
        binding_entity::Entity::find_by_id(allocation_id.to_owned())
            .one(&self.db)
            .await?
            .map(model_to_binding)
            .transpose()
    }

    pub async fn get_allocation(&self, id: &str) -> Result<Option<Allocation>> {
        let allocation = allocation_entity::Entity::find_by_id(id.to_owned())
            .one(&self.db)
            .await?;
        let Some(allocation) = allocation else {
            return Ok(None);
        };
        let binding = binding_entity::Entity::find_by_id(id.to_owned())
            .one(&self.db)
            .await?;
        Ok(Some(model_to_allocation(allocation, binding.as_ref())?))
    }

    pub async fn list_allocations(&self) -> Result<Vec<Allocation>> {
        let allocations = allocation_entity::Entity::find()
            .order_by_asc(allocation_entity::Column::Protocol)
            .order_by_asc(allocation_entity::Column::Port)
            .all(&self.db)
            .await?;
        let bindings = binding_entity::Entity::find().all(&self.db).await?;
        let bindings_by_id: HashMap<_, _> = bindings
            .iter()
            .map(|binding| (binding.allocation_id.as_str(), binding))
            .collect();

        allocations
            .into_iter()
            .map(|allocation| {
                let binding = bindings_by_id.get(allocation.id.as_str()).copied();
                model_to_allocation(allocation, binding)
            })
            .collect()
    }

    async fn setup_schema(&self) -> Result<()> {
        let builder = self.db.get_database_backend();
        let schema = Schema::new(builder);

        let mut allocations = schema.create_table_from_entity(allocation_entity::Entity);
        allocations.if_not_exists();
        self.db.execute(builder.build(&allocations)).await?;

        let mut bindings = schema.create_table_from_entity(binding_entity::Entity);
        bindings.if_not_exists();
        self.db.execute(builder.build(&bindings)).await?;

        let index = sea_orm::sea_query::Index::create()
            .if_not_exists()
            .name("idx_allocations_protocol_port")
            .table(allocation_entity::Entity)
            .col(allocation_entity::Column::Protocol)
            .col(allocation_entity::Column::Port)
            .unique()
            .to_owned();
        self.db.execute(builder.build(&index)).await?;
        Ok(())
    }

    async fn migrate_legacy_bindings(&self) -> Result<()> {
        self.db
            .execute(Statement::from_string(
                DbBackend::Sqlite,
                "INSERT INTO bindings(allocation_id, target_port, host, created_at_ms, updated_at_ms)
                 SELECT id, target_port, host, created_at_ms, updated_at_ms FROM allocations
                 WHERE target_port > 0
                   AND NOT EXISTS (SELECT 1 FROM bindings WHERE bindings.allocation_id = allocations.id);"
                    .to_owned(),
            ))
            .await?;
        Ok(())
    }

    async fn update_legacy_binding_columns(
        &self,
        allocation_id: &str,
        target_port: u16,
        host: Option<String>,
        updated_at_ms: i64,
    ) -> Result<()> {
        if let Some(model) = allocation_entity::Entity::find_by_id(allocation_id.to_owned())
            .one(&self.db)
            .await?
        {
            let mut active = model.into_active_model();
            active.target_port = Set(i32::from(target_port));
            active.host = Set(host);
            active.updated_at_ms = Set(updated_at_ms);
            active.update(&self.db).await?;
        }
        Ok(())
    }

    async fn clear_legacy_binding_columns(
        &self,
        allocation_id: &str,
        updated_at_ms: i64,
    ) -> Result<()> {
        if let Some(model) = allocation_entity::Entity::find_by_id(allocation_id.to_owned())
            .one(&self.db)
            .await?
        {
            let mut active = model.into_active_model();
            active.target_port = Set(0);
            active.host = Set(None);
            active.updated_at_ms = Set(updated_at_ms);
            active.update(&self.db).await?;
        }
        Ok(())
    }

    #[cfg(test)]
    async fn test_i64(&self, sql: &str, id: &str) -> i64 {
        sqlx::query_scalar(sql)
            .bind(id)
            .fetch_one(&self.pool)
            .await
            .unwrap()
    }

    #[cfg(test)]
    async fn test_optional_string(&self, sql: &str, id: &str) -> Option<String> {
        sqlx::query_scalar(sql)
            .bind(id)
            .fetch_one(&self.pool)
            .await
            .unwrap()
    }

    #[cfg(test)]
    async fn journal_mode(&self) -> String {
        sqlx::query_scalar("PRAGMA journal_mode;")
            .fetch_one(&self.pool)
            .await
            .unwrap()
    }

    #[cfg(test)]
    async fn busy_timeout_ms(&self) -> i64 {
        sqlx::query_scalar("PRAGMA busy_timeout;")
            .fetch_one(&self.pool)
            .await
            .unwrap()
    }
}

fn model_to_allocation(
    allocation: allocation_entity::Model,
    binding: Option<&binding_entity::Model>,
) -> Result<Allocation> {
    let protocol = Protocol::from_str(&allocation.protocol)?;
    let port = port_from_i32("port", allocation.port)?;
    let target_port = match binding {
        Some(binding) => Some(port_from_i32("binding.target_port", binding.target_port)?),
        None if allocation.target_port > 0 => {
            Some(port_from_i32("target_port", allocation.target_port)?)
        }
        None => None,
    };
    let host = binding
        .and_then(|binding| binding.host.clone())
        .or(allocation.host);
    Ok(Allocation {
        id: allocation.id,
        protocol,
        port,
        target_port,
        host,
        created_at_ms: allocation.created_at_ms,
        updated_at_ms: allocation.updated_at_ms,
    })
}

fn model_to_binding(binding: binding_entity::Model) -> Result<Binding> {
    Ok(Binding {
        allocation_id: binding.allocation_id,
        target_port: port_from_i32("binding.target_port", binding.target_port)?,
        host: binding.host,
        created_at_ms: binding.created_at_ms,
        updated_at_ms: binding.updated_at_ms,
    })
}

fn port_from_i32(field: &'static str, value: i32) -> Result<u16> {
    if value < 0 {
        return Err(RepositoryError::InvalidPort { field, value });
    }
    u16::try_from(value).map_err(Into::into)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::model::{Allocation, Binding, Protocol};
    use tempfile::NamedTempFile;

    async fn temp_repo() -> Repository {
        let file = NamedTempFile::new().unwrap();
        Repository::open(file.path()).await.unwrap()
    }

    fn allocation(
        id: &str,
        protocol: Protocol,
        port: u16,
        target_port: Option<u16>,
        host: Option<&str>,
    ) -> Allocation {
        Allocation {
            id: id.to_owned(),
            protocol,
            port,
            target_port,
            host: host.map(str::to_owned),
            created_at_ms: 1000,
            updated_at_ms: 1000,
        }
    }

    #[tokio::test]
    async fn sqlite_self_check_wal_and_busy_timeout_succeed() {
        let repo = temp_repo().await;
        repo.self_check().await.unwrap();
        assert_eq!(repo.journal_mode().await.to_ascii_lowercase(), "wal");
        assert_eq!(repo.busy_timeout_ms().await, 5000);
    }

    #[tokio::test]
    async fn sqlite_persists_and_reloads_allocations_and_bindings() {
        let repo = temp_repo().await;
        let alloc = allocation("a1", Protocol::Tcp, 10000, None, None);
        repo.insert_allocation(&alloc).await.unwrap();
        assert_eq!(repo.get_allocation("a1").await.unwrap().unwrap(), alloc);

        let binding = Binding {
            allocation_id: "a1".to_owned(),
            target_port: 8080,
            host: Some("127.0.0.1".to_owned()),
            created_at_ms: 1100,
            updated_at_ms: 1100,
        };
        repo.put_binding(&binding).await.unwrap();
        assert_eq!(repo.get_binding("a1").await.unwrap().unwrap(), binding);
        assert_eq!(
            repo.test_i64("SELECT target_port FROM allocations WHERE id = ?", "a1")
                .await,
            8080
        );
        assert_eq!(
            repo.test_i64("SELECT updated_at_ms FROM allocations WHERE id = ?", "a1")
                .await,
            1100
        );
        assert_eq!(
            repo.test_optional_string("SELECT host FROM allocations WHERE id = ?", "a1")
                .await
                .as_deref(),
            Some("127.0.0.1")
        );
        let hydrated = repo.get_allocation("a1").await.unwrap().unwrap();
        assert_eq!(hydrated.target_port, Some(8080));
        assert_eq!(hydrated.host.as_deref(), Some("127.0.0.1"));

        let updated_binding = Binding {
            allocation_id: "a1".to_owned(),
            target_port: 9090,
            host: None,
            created_at_ms: 1100,
            updated_at_ms: 1150,
        };
        repo.put_binding(&updated_binding).await.unwrap();
        assert_eq!(
            repo.get_binding("a1").await.unwrap().unwrap(),
            updated_binding
        );
        assert_eq!(
            repo.test_i64("SELECT target_port FROM allocations WHERE id = ?", "a1")
                .await,
            9090
        );
        assert_eq!(
            repo.test_optional_string("SELECT host FROM allocations WHERE id = ?", "a1")
                .await,
            None
        );
        assert_eq!(
            repo.test_i64("SELECT updated_at_ms FROM allocations WHERE id = ?", "a1")
                .await,
            1150
        );

        assert!(repo.delete_binding("a1", 1200).await.unwrap());
        assert_eq!(
            repo.test_i64("SELECT target_port FROM allocations WHERE id = ?", "a1")
                .await,
            0
        );
        assert_eq!(
            repo.test_i64("SELECT updated_at_ms FROM allocations WHERE id = ?", "a1")
                .await,
            1200
        );
        assert_eq!(
            repo.test_optional_string("SELECT host FROM allocations WHERE id = ?", "a1")
                .await,
            None
        );
        let cleared = repo.get_allocation("a1").await.unwrap().unwrap();
        assert_eq!(cleared.target_port, None);
        assert_eq!(cleared.host, None);
        assert!(!repo.delete_binding("a1", 1300).await.unwrap());

        repo.put_binding(&binding).await.unwrap();
        assert!(repo.delete_allocation("a1").await.unwrap());
        assert!(repo.get_allocation("a1").await.unwrap().is_none());
        assert!(repo.get_binding("a1").await.unwrap().is_none());
        assert!(!repo.delete_allocation("a1").await.unwrap());
    }

    #[tokio::test]
    async fn sqlite_lists_allocations_ordered_by_protocol_then_port() {
        let repo = temp_repo().await;
        repo.insert_allocation(&allocation("udp", Protocol::Udp, 10002, None, None))
            .await
            .unwrap();
        repo.insert_allocation(&allocation("tcp", Protocol::Tcp, 10001, None, None))
            .await
            .unwrap();
        repo.insert_allocation(&allocation("both", Protocol::Both, 10000, None, None))
            .await
            .unwrap();
        let ids: Vec<String> = repo
            .list_allocations()
            .await
            .unwrap()
            .into_iter()
            .map(|row| row.id)
            .collect();
        assert_eq!(ids, vec!["both", "tcp", "udp"]);
    }

    #[tokio::test]
    async fn sqlite_migrates_legacy_allocation_binding_into_bindings_table() {
        let file = NamedTempFile::new().unwrap();
        {
            let pool = SqlitePoolOptions::new()
                .max_connections(1)
                .connect_with(
                    SqliteConnectOptions::new()
                        .filename(file.path())
                        .create_if_missing(true),
                )
                .await
                .unwrap();
            sqlx::query(
                "CREATE TABLE allocations (
                    id TEXT PRIMARY KEY,
                    protocol TEXT NOT NULL,
                    port INTEGER NOT NULL,
                    target_port INTEGER NOT NULL,
                    host TEXT,
                    created_at_ms INTEGER NOT NULL,
                    updated_at_ms INTEGER NOT NULL,
                    UNIQUE(protocol, port)
                );",
            )
            .execute(&pool)
            .await
            .unwrap();
            sqlx::query(
                "INSERT INTO allocations(id, protocol, port, target_port, host, created_at_ms, updated_at_ms)
                 VALUES('legacy', 'tcp', 10000, 8080, '127.0.0.1', 10, 20);",
            )
            .execute(&pool)
            .await
            .unwrap();
            pool.close().await;
        }
        let repo = Repository::open(file.path()).await.unwrap();
        let binding = repo.get_binding("legacy").await.unwrap().unwrap();
        assert_eq!(binding.target_port, 8080);
        assert_eq!(binding.host.as_deref(), Some("127.0.0.1"));
        assert_eq!(binding.created_at_ms, 10);
        assert_eq!(binding.updated_at_ms, 20);
    }

    #[tokio::test]
    async fn sqlite_delete_allocation_removes_existing_binding_row() {
        let repo = temp_repo().await;
        repo.insert_allocation(&allocation("a1", Protocol::Tcp, 10000, None, None))
            .await
            .unwrap();
        repo.put_binding(&Binding {
            allocation_id: "a1".to_owned(),
            target_port: 8080,
            host: Some("127.0.0.1".to_owned()),
            created_at_ms: 1100,
            updated_at_ms: 1100,
        })
        .await
        .unwrap();

        assert!(repo.delete_allocation("a1").await.unwrap());

        assert_eq!(
            repo.test_i64(
                "SELECT COUNT(*) FROM bindings WHERE allocation_id = ?",
                "a1"
            )
            .await,
            0
        );
        assert!(repo.get_binding("a1").await.unwrap().is_none());
    }
}
